### PR TITLE
Fixes for non-inverted relay

### PR DESCRIPTION
**Pull request by NDR008**
I created a new turn_on_pin function
and called that function when print_started.

~~Also I made sure print_started is only called when that channel is active, because some people might have devices connected to relays that they de-activated.~~

### CODE REVIEW

**Comment by NDR008** [comment]
@borisbu I think this fixes the bug

**Comment by NDR008** [comment]
This change is not merged yet. Did you install my branch?

**Comment by storbi81** [comment]
I installed everything via the plugin-manager from Octoprint.

**Comment by NDR008** [comment]
> Good morning!
> 
> I think this change is responsible for a new problem that I have now with my printer. On print_started, both of my relais switched to on now, but I can't change this anywhere.
> 
> A few days ago, i tried to install the plugin Octoprint-Smart-Filament-Sensor from Anni L., but it will not work with OctoRelay (only the standard R1-R4 will be displayed in Navbar, nothing worked, you can configure what you want, no change). I've deinstalled both of the plugins and reinstalled OctoRelay again. I think, now I have some trouble with the different versions.

This change is not merged yet. Did you install my branch? How did you confirm that this change caused the issue?
What do you mean multiple-versions?

**Comment by NDR008** [comment]
> I installed everything via the plugin-manager from Octoprint.

Then you did not use this change, because this change is not yet part of the publicly available plugin from the plugin manager.

**Comment by NDR008** [comment]
> I created a new turn_on_pin function and called that function when print_started. Also I made sure print_started is only called when that channel is active, because some people might have devices connected to relays that they de-activated.

Thinking about your problem, I think this problem is actually one of the things I fixed in my change.
Are you able to load my code to test it? you have to manually download my python file and substitute it on your OctoPi/Octoprint.

**Comment by ArmEagle** [comment]
Thank you. This fixed the exact problem you described in #46 that I was having myself too after adding a relay to control the power to the printer. And luckily it's very easy to add a modified plugin to OctoPi.

**Comment by Zbysekz** [comment]
Confirmed - this solves problem when controlling printer and lights by GPIO outputs normally works, but if you start printing it shuts down the printer. Thanks!

**Comment by RobinTail** [comment]
@ArmEagle && @Zbysekz && @NDR008 ,

do I understand it correctly, that this PR fixes the issue #46 ?

**Comment by NDR008** [comment]
In my opinion yes.  I've been using it over a year.

File: octoprint_octorelay/__init__.py [hunk 1]
```diff
@@ -349,6 +346,16 @@ def turn_off_pin(self, relay_pin, inverted, cmdOFF):
         self._logger.info("pin: {} turned off".format(relay_pin))
         self.update_ui()
 
+    def turn_on_pin(self, relay_pin, inverted, cmdON):
+        GPIO.setup(relay_pin, GPIO.OUT)
```
Comment (RobinTail) [inline octoprint_octorelay/__init__.py:350]: @NDR008 , is it ok, that exactly the same command is being executed right before calling this function, in line 318?

https://github.com/borisbu/OctoRelay/blob/cd52345b03b61de12cfd26eb275bf42d0441f47c/octoprint_octorelay/__init__.py#L318

File: octoprint_octorelay/__init__.py [hunk 2]
```diff
@@ -349,6 +346,16 @@ def turn_off_pin(self, relay_pin, inverted, cmdOFF):
...
+        # XOR with inverted
+        GPIO.output(relay_pin, inverted == False)
+        GPIO.setwarnings(True)
+        if cmdON:
+            os.system(cmdON)
+        self._logger.info("pin: {} turned on".format(relay_pin))
+        self.update_ui()
```
Comment (RobinTail) [inline octoprint_octorelay/__init__.py:357]: @NDR008 , the function `turn_on_pin` is being called within a cycle introduced in line 309:

https://github.com/borisbu/OctoRelay/blob/cd52345b03b61de12cfd26eb275bf42d0441f47c/octoprint_octorelay/__init__.py#L309

Which makes this statement (updating the UI) being fired for each of the relays in configuration.

Apart from that, the same UI update is fired in line 320 anyway, after finishing the cycle:

https://github.com/borisbu/OctoRelay/blob/cd52345b03b61de12cfd26eb275bf42d0441f47c/octoprint_octorelay/__init__.py#L320

Question: is there any benefit of firing the UI update within the cycle? Can we keep the single UI update in the end of the cycle?

**Review by RobinTail — CHANGES_REQUESTED** [review]
We've recently solved #55 that filters out deactivated relays as well.
Please adjust you implementation by resolving the conflicts, @NDR008

**Comment by RobinTail** [comment]
Conflict resolved by preserving the solution on only active relays handling in target branch that was fixed in another PR: #55.

File: octoprint_octorelay/__init__.py (hunk 2 above)
Comment (RobinTail) [reply octoprint_octorelay/__init__.py:359]:
```suggestion
```

removing UI update here, since it's called in the end of `print_stopped` which calls this method in a cycle

File: octoprint_octorelay/__init__.py [hunk 3]
```diff
@@ -313,12 +313,12 @@ def print_started(self):
             relay_pin = int(settings["relay_pin"])
             inverted = settings['inverted_output']
             autoONforPrint = settings['autoONforPrint']
+            cmdON = settings['cmdON']
             active = settings["active"]
             if autoONforPrint and active:
                 self._logger.debug("turning on pin: {}, index: {}".format(relay_pin, index))
                 GPIO.setup(relay_pin, GPIO.OUT)
```
Comment (RobinTail) [inline octoprint_octorelay/__init__.py:320]:
```suggestion
```

removing this line here, since:
- same one exists in the `turn_on_pin` which is called in the next line
- for consistency with `turn_off_pin` method doing the same

File: octoprint_octorelay/__init__.py (hunk 1 above)
Comment (RobinTail) [reply octoprint_octorelay/__init__.py:352]: made a suggestion to keep only this one

**Review by RobinTail — COMMENTED** [review]
OK. Now I like its appearance.

**Comment by RobinTail** [comment]
The commands I used to update this PR based on the latest changes:
```
gh pr checkout 48
git merge develop
git push git@github.com:NDR008/OctoRelay.git HEAD:master
```

**Comment by RobinTail** [comment]
All tests succeeded, but `os.system` has to be mocked for `test_print_started`:

Screenshot 2023-07-23 at 12 29 41

**Comment by RobinTail** [comment]
Sweet!

Screenshot 2023-07-23 at 13 37 27

File: octoprint_octorelay/__init__.py [hunk 4]
```diff
@@ -333,8 +332,6 @@ def print_stopped(self):
             cmdOFF = settings['cmdOFF']
             active = settings["active"]
             if autoOFFforPrint and active:
-                self._logger.debug("turn off pin: {} in {} seconds. index: {}".format(
-                    relay_pin, autoOffDelay, index))
```
Comment (RobinTail) [inline octoprint_octorelay/__init__.py:337]: not sure why this was removed.
it does not seem to be related to the fix

File: octoprint_octorelay/test__init.py [hunk 5]
```diff
@@ -307,6 +307,20 @@ def test_turn_off_pin(self, systemMock):
             systemMock.assert_called_with("CommandMock")
         restoreUI()
 
+    @patch('os.system')
+    def test_turn_off_on(self, systemMock):
```
Comment (RobinTail) [inline octoprint_octorelay/test__init.py:311]: wrong naming

```suggestion
    def test_turn_on_pin(self, systemMock):
```

**Review by RobinTail — APPROVED** [review]
👍🏽

**Comment by RobinTail** [comment]
QA passed
